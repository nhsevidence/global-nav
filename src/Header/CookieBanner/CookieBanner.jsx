import React, { Component } from "react";
import Cookies from "js-cookie";

import styles from "./CookieBanner.module.scss";

export const CookieName = "seen_cookie_message";

// Increment this and release a new version to force users to re-accept the message
export const CookieMessageVersion = 2;

export default class CookieBanner extends Component {
	constructor(props) {
		super(props);

		const cookieValue = Cookies.getJSON(CookieName);

		// "yes" was used for the previous version of TopHat
		if (!cookieValue || cookieValue === "yes") {
			//  Delete old cookie by setting expiry date in the past
			Cookies.remove(CookieName);
			this.state = {
				hasSeenPreviousVersion: false,
				isClosed: false
			};
		} else {
			const seenVersion = cookieValue;

			this.state = {
				hasSeenPreviousVersion: seenVersion < CookieMessageVersion,
				isClosed: seenVersion === CookieMessageVersion
			};
		}

		this.handleClick = this.handleClick.bind(this);
	}

	getCookieDomain(host) {
		return ~host.indexOf("nice.org.uk") ? "nice.org.uk" : host;
	}

	handleClick() {
		const cookieOptions = {
			secure: false,
			expires: 365 // In days
		};

		if (window.location.hostname !== "localhost") {
			cookieOptions.domain = this.getCookieDomain(window.location.hostname);
		}

		Cookies.set(CookieName, CookieMessageVersion, cookieOptions);

		this.setState({
			isClosed: true
		});
	}

	render() {
		if (this.state.isClosed) return null;

		return (
			<div
				className={styles.banner}
				role="dialog"
				aria-labelledby="cookie-banner-title"
				aria-describedby="cookie-banner-description"
			>
				<div className={styles.container}>
					<h2 className={styles.heading} id="cookie-banner-title">
						Your privacy
					</h2>
					<p className={styles.message} id="cookie-banner-description">
						{this.state.hasSeenPreviousVersion
							? "Our cookie policy has changed since you last accepted it. "
							: "We have placed cookies on your device to help make this website better. "}
						By continuing to use our website, you agree to this. You can find
						out more information and how to change some cookie settings in our{" "}
						<a href="https://www.nice.org.uk/terms-and-conditions#cookies">
							cookie&nbsp;policy
						</a>
						.
					</p>
					<button
						className={styles.button}
						type="button"
						onClick={this.handleClick}
					>
						Accept and close
					</button>
				</div>
			</div>
		);
	}
}
