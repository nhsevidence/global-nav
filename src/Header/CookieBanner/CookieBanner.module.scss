@import '~@nice-digital/nds-core/scss/core';

.banner {
  background: $nds-colour-nice-grey-3;
  border-bottom: 1px solid $nds-colour-border;
  padding: rem($nds-spacing-medium 0);
  text-align: left;
}

.container {
  @include nds-container;
}

.heading {
  @include font(h5);
  font-weight: bold;
  margin-top: 0;
}

.message {
  @include p;

  a {
    @include link-default;
  }
}

.button {
  appearance: none;
  background: $nds-colour-btn-primary;
  border: em(1) solid $nds-colour-btn-primary-text;
  border-radius: 0;
  color: $nds-colour-btn-primary-text;
  cursor: pointer;
  display: inline-block;
  margin: 0;
  padding: em($nds-spacing-small $nds-spacing-medium);
  position: relative;
  text-align: left;
  text-decoration: none !important; // sass-lint:disable-line no-important
  vertical-align: top;
  white-space: nowrap;

  &:focus {
    background: $nds-colour-btn-primary;
    border-color: $nds-colour-focus-inverse;
    color: $nds-colour-btn-primary-text;
    outline: 3px solid $nds-colour-focus;
  }
}
