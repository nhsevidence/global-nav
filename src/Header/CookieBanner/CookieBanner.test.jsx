import React from "react";
import Cookies from "js-cookie";
import CookieBanner, { CookieMessageVersion, CookieName } from "./CookieBanner";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

describe("CookieBanner", () => {
	const oldCookie = document.cookie;

	beforeEach(() => {
		// Reset overridden cookie property
		document.cookie = oldCookie;
		Cookies.set(CookieName, null);
	});

	it("Renders without crashing", () => {
		const wrapper = shallow(<CookieBanner />);
		expect(wrapper).toHaveLength(1);
	});

	it("Doesn't render by default/on the server", () => {
		const wrapper = shallow(<CookieBanner />, {
			disableLifecycleMethods: true
		});

		expect(wrapper.isEmptyRender()).toBe(true);
	});

	it("Renders after rehydration on the client", () => {
		const wrapper = shallow(<CookieBanner />);

		wrapper.update();

		expect(wrapper.isEmptyRender()).toBe(false);
	});

	it("Matches snapshot for new visitor", () => {
		const wrapper = shallow(<CookieBanner />);

		wrapper.update();

		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it("Matches snapshot for old cookie value", () => {
		Cookies.set(CookieName, "yes");

		const wrapper = shallow(<CookieBanner />);

		wrapper.update();

		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it("Hides cookie message when seen correct message version", () => {
		Cookies.set(CookieName, CookieMessageVersion);

		const wrapper = shallow(<CookieBanner />);

		wrapper.update();

		expect(wrapper.isEmptyRender()).toBe(true);
	});

	it("Matches snapshot when seen old message version", () => {
		Cookies.set(CookieName, CookieMessageVersion - 1);

		const wrapper = shallow(<CookieBanner />);

		wrapper.update();

		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it("Triggers click handler on close button click", () => {
		const handleClick = jest.spyOn(CookieBanner.prototype, "handleClick");
		const wrapper = shallow(<CookieBanner />);

		wrapper.update();

		wrapper
			.find("button")
			.at(0)
			.simulate("click");

		expect(handleClick).toHaveBeenCalledTimes(1);
	});

	it("Hides banner on close button click", () => {
		const wrapper = shallow(<CookieBanner />);

		wrapper.update();

		wrapper
			.find("button")
			.at(0)
			.simulate("click");

		expect(wrapper.isEmptyRender()).toBe(true);
	});

	it("Stores cookie for 365 days on niceorg domain on button click", () => {
		const cookieSet = jest.fn();

		Object.defineProperty(document, "cookie", {
			get: jest.fn().mockImplementation(() => null),
			set: cookieSet,
			configurable: true
		});

		const wrapper = shallow(<CookieBanner />);

		wrapper.update();

		wrapper
			.find("button")
			.at(0)
			.simulate("click");

		expect(cookieSet).toHaveBeenCalledTimes(2);

		const date = new Date();
		date.setDate(date.getDate() + 365);

		expect(cookieSet.mock.calls[1][0]).toEqual(
			`${CookieName}=${CookieMessageVersion}; path=/; expires=${date.toUTCString()}; domain=nice.org.uk`
		);

		const pastDate = new Date();
		pastDate.setDate(pastDate.getDate() - 1);
		pastDate.setHours(pastDate.getHours() - 1); // TODO: revisit this for BST!

		expect(cookieSet.mock.calls[0][0]).toEqual(
			`${CookieName}=; path=/; expires=${pastDate.toUTCString()}`
		);
	});

	it("Stores cookie on evidence domain on button click", () => {
		const oldLocation = global.window.location;
		delete global.window.location;
		global.window.location = new URL("https://www.evidence.nhs.uk/");

		const cookieSet = jest.fn();

		Object.defineProperty(document, "cookie", {
			get: jest.fn().mockImplementation(() => null),
			set: cookieSet,
			configurable: true
		});

		const wrapper = shallow(<CookieBanner />);

		wrapper.update();

		wrapper
			.find("button")
			.at(0)
			.simulate("click");

		const date = new Date();
		date.setDate(date.getDate() + 365);

		expect(cookieSet.mock.calls[1][0]).toEqual(
			`${CookieName}=${CookieMessageVersion}; path=/; expires=${date.toUTCString()}; domain=www.evidence.nhs.uk`
		);

		// Tidy up
		global.window.location = oldLocation;
	});
});
