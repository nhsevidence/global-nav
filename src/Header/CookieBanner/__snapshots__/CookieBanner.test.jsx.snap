// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookieBanner Matches snapshot for new visitor 1`] = `
<div
  aria-describedby="cookie-banner-description"
  aria-labelledby="cookie-banner-title"
  className="banner"
  role="dialog"
>
  <div
    className="container"
  >
    <h2
      className="heading"
      id="cookie-banner-title"
    >
      Your privacy
    </h2>
    <p
      className="message"
      id="cookie-banner-description"
    >
      We have placed cookies on your device to help make this website better. 
      You can find out more information and how to change some cookie settings in our
       
      <a
        href="https://www.nice.org.uk/terms-and-conditions#cookies"
      >
        cookie policy
      </a>
      .
    </p>
    <button
      className="button"
      onClick={[Function]}
      type="button"
    >
      Accept and close
    </button>
  </div>
</div>
`;

exports[`CookieBanner Matches snapshot for old cookie value 1`] = `
<div
  aria-describedby="cookie-banner-description"
  aria-labelledby="cookie-banner-title"
  className="banner"
  role="dialog"
>
  <div
    className="container"
  >
    <h2
      className="heading"
      id="cookie-banner-title"
    >
      Your privacy
    </h2>
    <p
      className="message"
      id="cookie-banner-description"
    >
      We have placed cookies on your device to help make this website better. 
      You can find out more information and how to change some cookie settings in our
       
      <a
        href="https://www.nice.org.uk/terms-and-conditions#cookies"
      >
        cookie policy
      </a>
      .
    </p>
    <button
      className="button"
      onClick={[Function]}
      type="button"
    >
      Accept and close
    </button>
  </div>
</div>
`;

exports[`CookieBanner Matches snapshot when seen old message version 1`] = `
<div
  aria-describedby="cookie-banner-description"
  aria-labelledby="cookie-banner-title"
  className="banner"
  role="dialog"
>
  <div
    className="container"
  >
    <h2
      className="heading"
      id="cookie-banner-title"
    >
      Your privacy
    </h2>
    <p
      className="message"
      id="cookie-banner-description"
    >
      Our cookie policy has changed since you last accepted it. 
      You can find out more information and how to change some cookie settings in our
       
      <a
        href="https://www.nice.org.uk/terms-and-conditions#cookies"
      >
        cookie policy
      </a>
      .
    </p>
    <button
      className="button"
      onClick={[Function]}
      type="button"
    >
      Accept and close
    </button>
  </div>
</div>
`;
